### DISCUSSION TITLE
chroot SystemsD project

### DISCUSSION THREAD

**Discussion by TomTravis** [dnschneid/crouton#4859] [Q&A] [answered]
**New project task and invites have been sent**
![Screenshot 2022-11-05 10 09 40 AM](https://user-images.githubusercontent.com/46386261/200132555-3c26585c-d27f-42b9-b1dc-48f9469d1b27.png)

it starts with the "enter-chroot" script
this is need work and testing

Init.D and the rc* dirs in /etc do not start on enter-chroot

I m using monit for testing it certainty works just fine
[mmonit.com](https://mmonit.com/)

The latest systemd is on Github they are very active, Crostini has systemd now i believe bullseye R

# Start the chroot and any specified command
if [ -n "$NOLOGIN" ]; then
    env -i TERM="$TERM" chroot "$CHROOT" "$@" || ret=$?
    # Handle the return value when running the setup script.
    if [ "$NOLOGIN" = 2 ]; then
        # If it succeeded yet still exists, run it again.
        if [ "$ret" = 0 -a -f "$CHROOT$SETUPSCRIPT" ]; then
            sh -e "$BINDIR/unmount-chroot" -yxc "$CHROOTS" -- "$NAME"
            # We don't need to return from this.
            exec sh -e "$BINDIR/enter-chroot" -c "$CHROOTS" -n "$NAME" -xx
        elif [ "$ret" != 0 ]; then
            error "$ret" 'Failed to complete chroot setup.'
        fi
    fi
else
    # Check and run rc.local
    if [ -n "$firstrun" -a -x "$CHROOT/etc/rc.local" ]; then
        ( sleep 10; echo "Still running chroot's rc.local..." 2>&1 ) &
        noticepid="$!"
        addtrap "kill '$noticepid' 2>/dev/null"
        chrootcmd 'exec /etc/rc.local >/dev/null 2>/dev/null </dev/null' \
            || ret=$?
        undotrap
        kill "$noticepid" 2>/dev/null || true
        wait "$noticepid" || true
        if [ "$ret" != 0 ]; then
            echo "WARNING: /etc/rc.local failed with code $ret" 1>&2
            ret=0
        fi
    fi

    if [ $# = 0 -o -n "$LOGIN" ]; then
        env -i TERM="$TERM" chroot "$CHROOT" "$@" su - "$USERNAME" || ret=$?
    else
        # Escape out the command
        cmd="export SHELL='$CHROOTSHELL';"
        if [ -n "$TMPXMETHOD" ]; then
            cmd="$cmd export XMETHOD='$TMPXMETHOD';"
        fi
        for param in "$@"; do
            cmd="$cmd'`echo -n "$param" | sed "s/'/'\\\\\\''/g"`' "
        done
        env -i TERM="$TERM" chroot "$CHROOT" \
            su -s '/bin/sh' -c "$cmd" - "$USERNAME" \
            || ret=$?
    fi
fi

 # Launch systemd-logind if available and not already running
    # Whitelisted for saucy and trusty
    systemd_dir="`fixabslinks '/run/systemd'`"
    if [ -x "$CHROOT/lib/systemd/systemd-logind" ] && \
            [ ! -d "$systemd_dir" ] && \
            [ "$CHROOTRELEASE" = 'saucy' -o "$CHROOTRELEASE" = 'trusty' ]; then
        # Every piece of systemd code ever assumes that this directory exists
        mkdir -p "$systemd_dir"

        # Create systemd cgroup if necessary
        if ! mountpoint -q "$CHROOT/sys/fs/cgroup/systemd"; then
            mkdir -p "$CHROOT/sys/fs/cgroup/systemd"
            mount -t cgroup -o nosuid,noexec,nodev,none,name=systemd systemd \
                  "$CHROOT/sys/fs/cgroup/systemd"
        fi
        # systemd-logind doesn't fork
        chrootcmd "/lib/systemd/systemd-logind >/dev/null 2>&1 </dev/null &"

https://github.dev/dnschneid/crouton/blob/master/host-bin/enter-chroot

the new github ediitor is vscode very cool
    fi

**Comment by dnschneid** [comment]
systemd support basically requires crouton to implement namespacing, which is way out of the scope of "maintenance".  There was an attempt in #2474 if you want something to start from.

**Reply by TomTravis** [reply]
![Screenshot from 2022-11-03 09-45-48](https://user-images.githubusercontent.com/46386261/199784903-edd9adc3-6515-4f89-8c11-08fc27497701.png)
the  User IS is certainly tracked
https://github.com/systemd/systemd

yes this can be done in my fork for those who are interested Xorg uses systemd , so it is kinda maintenance it local work for me

the startup process need to start systemD and crostini has does it. if namespaceing is required systemd will have it as part of the install 

basically requires crouton to implement namespacing, well that was 2016 nft tables are also required it not required for startup

**Comment by TomTravis** [comment]
REQUIREMENTS:
        Linux kernel ≥ 3.15
                     ≥ 4.3 for ambient capabilities
                     ≥ 4.5 for pids controller in cgroup v2
                     ≥ 4.6 for cgroup namespaces
                     ≥ 4.9 for RENAME_NOREPLACE support in vfat
                     ≥ 4.10 for cgroup-bpf egress and ingress hooks
                     ≥ 4.15 for cgroup-bpf device hook and cpu controller in cgroup v2
                     ≥ 4.17 for cgroup-bpf socket address hooks
                     ≥ 4.20 for PSI (used by systemd-oomd)
                     ≥ 5.3 for bounded loops in BPF program
                     ≥ 5.4 for signed Verity images
                     ≥ 5.7 for BPF links and the BPF L

i using for integration of systemD chroot
PRETTY_NAME="Kali GNU/Linux Rolling"
NAME="Kali GNU/Linux"
ID=kali
VERSION="2022.3"
VERSION_ID="2022.3"
VERSION_CODENAME="kali-rolling"
ID_LIKE=debian
ANSI_COLOR="1;31"
HOME_URL="https://www.kali.org/"
SUPPORT_URL="https://forums.kali.org/"
BUG_REPORT_URL="https://bugs.kali.org/"

**Comment by TomTravis** [comment]
hello the lastest systemD was cloned on Kali and is buliding 
sudo apt build -dep systemd
and jinja2 python had to me installed for the build 

systemd is super configurable and can be setup for many envs 

the idea is to start the Demon via enter-chroot and see what happens

**Comment by TomTravis** [comment]
great news no error chroot cannot run
(kr)z@localhost:~$ /usr/bin/systemd --test --system
systemd 251~rc3-2927-g4c72e71 running in test system mode (+PAM +AUDIT +SELINUX>
Detected architecture arm64.
Cannot determine cgroup we are running in: No medium found
Failed to allocate manager object: No medium found

now need the startup commands

The Tragedy of systemd
[https://youtu.be/o_AIw9bGogo](url)
you need to Cut and paste the link to watch the video it somewhat useful 2019

**Comment by TomTravis** [comment]
moving Forward now some sort of schedule and task can be written
dbus[6980]: Unknown username "systemd-timesync" in message bus configuration file
dbus[6980]: Unknown username "systemd-oom" in message bus configuration file
dbus[6980]: Unknown username "systemd-resolve" in message bus configuration file
the idea to start systemd via init is next, this can manage the mounts PTS in chroot will see

**Comment by TomTravis** [comment]
this is the project start
if you wish to join please do
https://github.com/users/TomTravis/projects/3/views/1

**Comment by TomTravis** [comment]
[PTS_code.txt](https://github.com/dnschneid/crouton/files/9940356/PTS_code.txt)
this is in work enter-chroot update systemd startup

this has taken a day is so sysVinit does not start yet so this is the current work maybe to use monit or supervisor 

have yet to review upstart or rclocal  a review of the crostini seems like a great idea more like a container chroot debootstrap only get you so far this also can be reviewed IE update backports and nvidia

**Comment by TomTravis** [comment, marked as answer]
![Screenshot 2022-11-07 7 24 36 PM](https://user-images.githubusercontent.com/46386261/200479488-a9618fd6-0712-4566-927a-3f8fcb4e3554.png)
move to chroot systemd project
